atlasdb-backup/src/main/java/com/palantir/atlasdb/backup/CoordinationServiceRecorder.java

```diff
@@ -0,0 +1,86 @@
+import com.palantir.logsafe.SafeArg;
+import com.palantir.logsafe.logger.SafeLogger;
+import com.palantir.logsafe.logger.SafeLoggerFactory;
+import java.util.Optional;
+import java.util.function.Function;
+
+final class CoordinationServiceRecorder {
+    private static final SafeLogger log = SafeLoggerFactory.get(CoordinationServiceRecorder.class);
+
+    private final Function<Namespace, KeyValueService> keyValueServiceFactory;
+    private final BackupPersister backupPersister;
+
+    CoordinationServiceRecorder(
+            Function<Namespace, KeyValueService> keyValueServiceFactory, BackupPersister backupPersister) {
+        this.keyValueServiceFactory = keyValueServiceFactory;
+        this.backupPersister = backupPersister;
+    }
+
+    public void storeFastForwardState(CompletedBackup completedBackup) {
+        Namespace namespace = completedBackup.getNamespace();
+        Optional<InternalSchemaMetadataState> maybeMetadata =
+                fetchSchemaMetadata(namespace, completedBackup.getBackupEndTimestamp());
+
+        maybeMetadata.ifPresentOrElse(
+                metadata -> backupPersister.storeSchemaMetadata(completedBackup.getBackupId(), namespace, metadata),
+                () -> logEmptyMetadata(namespace));
+    }
+
+    private void logEmptyMetadata(Namespace namespace) {
+        log.info("No metadata stored in coordination service for namespace.", SafeArg.of("namespace", namespace));
+    }
+
+    private Optional<InternalSchemaMetadataState> fetchSchemaMetadata(Namespace namespace, long timestamp) {
+        try (KeyValueService kvs = keyValueServiceFactory.apply(namespace)) {
+            if (!kvs.getAllTableNames().contains(AtlasDbConstants.COORDINATION_TABLE)) {
+                return Optional.empty();
+            }
+            CoordinationService<InternalSchemaMetadata> coordination =
+                    CoordinationServices.createDefault(kvs, () -> timestamp, false);
+
+            return getValidMetadata(coordination, timestamp).map(InternalSchemaMetadataState::of);
+        }
+    }
+
+    private Optional<ValueAndBound<InternalSchemaMetadata>> getValidMetadata(
+            CoordinationService<InternalSchemaMetadata> coordination, long timestamp) {
+        return coordination.getValueForTimestamp(timestamp).or(() -> tryExtendValidityBound(coordination, timestamp));
+    }
+
+    private Optional<ValueAndBound<InternalSchemaMetadata>> tryExtendValidityBound(
+            CoordinationService<InternalSchemaMetadata> coordination, long timestamp) {
+        CheckAndSetResult<ValueAndBound<InternalSchemaMetadata>> casResult = coordination.tryTransformCurrentValue(
+                valueAndBound -> valueAndBound.value().orElseGet(InternalSchemaMetadata::defaultValue));
+        ValueAndBound<InternalSchemaMetadata> persisted = Iterables.getOnlyElement(casResult.existingValues());
+        return (persisted.bound() >= timestamp) ? Optional.of(persisted) : Optional.empty();
```

Comment:
We still have to keep retrying until this is successful, we just don't have to do it with a fresh timestamp, but we can't continue until the bound contains the FF timestamp

Reply:
I see, yeah I had thought that trying again with the same timestamp would simply yield the same result.